File: dotnet/src/Microsoft.Agents.Validation/Validators/MessageValidators/DeveloperMessageValidator.cs
using Microsoft.Agents.Xml.Generated.Models;

namespace Microsoft.Agents.Xml.Validation.Validators.MessageValidators;

/// <summary>
/// Validator for DeveloperMessage.
/// </summary>
public class DeveloperMessageValidator : MessageValidatorBase<DeveloperMessage>
{
    protected override List<ValidationError> ValidateSpecificFields(DeveloperMessage message, ValidationContext? context)
    {
        var errors = new List<ValidationError>();

        // MSG-009: DeveloperMessage content must be non-empty
        if (string.IsNullOrWhiteSpace(message.Content))
        {
            errors.Add(CreateError(
                ValidationErrorCode.MSG_009,
                "DeveloperMessage content must be non-empty",
                "Content"));
        }

        // ROLE-002: DeveloperMessage can only contain TextContent
        // Note: DeveloperMessage uses Content property, not Contents list

        return errors;
    }
}

File: typescript/packages/agents-protocol-validation/src/validators/message/DeveloperMessageValidator.ts
import type { DeveloperMessage } from '@microsoft/agents-protocol-abstractions';
import type { ValidationContext } from '../../core/ValidationContext';
import { MessageValidatorBase } from '../base/MessageValidatorBase';
import { ValidationError } from '../../core/ValidationError';
import { ValidationErrorCode } from '../../core/ValidationErrorCode';

/**
 * Validator for DeveloperMessage.
 */
export class DeveloperMessageValidator extends MessageValidatorBase<DeveloperMessage> {
  protected validateSpecificFields(message: DeveloperMessage, context?: ValidationContext): ValidationError[] {
    // MSG-009: Developer messages have no special fields beyond common ones
    return [];
  }

  protected getAllowedContentTypes(): string[] {
    // ROLE-002: Developer messages can only contain text, document
    return ['text', 'document'];
  }

  protected getRoleSpecificErrorCode(): string {
    return ValidationErrorCode.ROLE_002;
  }

  protected getMessageTypeName(): string {
    return 'DeveloperMessage';
  }

  protected requiresContent(): boolean {
    return false; // Developer messages can be empty
  }
}
